Replace OkHttp with Mutiny in Vault extension

vsevel opened the pull request:
This aims at removing the okhttp client used in the vault extension to interact with the vault server, and replace it with a vertx web client.
There was an earlier attempt in #8885, which this PR is initially based on.
If successful, this would fix #4162 as well.

/cc @sberyozkin @cescoffier @geoand

vsevel committed:
Replace OkHttp with Mutiny in Vault extension #14319

geoand commented:
I am actually not sure if this will work properly because the boostrap config recorders are executed before any runtime-init recorders so I don't know what that means for the vert.x instance

vsevel commented:
I am having an issue with `Vertx` being (unfortunately) a synthetic bean:
```
Caused by: javax.enterprise.inject.CreationException: Synthetic bean instance for io.vertx.core.Vertx not initialized yet: io_vertx_core_Vertx_97d170e1550eee4afc0af065b78cda302a97674c
        - a synthetic bean initialized during RUNTIME_INIT must not be accessed during STATIC_INIT
        - RUNTIME_INIT build steps that require access to synthetic beans initialized during RUNTIME_INIT should consume the SyntheticBeansRuntimeInitBuildItem
        at io.vertx.core.Vertx_7c31d5064e6d5db248fafa10f188ca6323f3f983_Synthetic_Bean.create(Vertx_7c31d5064e6d5db248fafa10f188ca6323f3f983_Synthetic_Bean.zig:117)
...
        at io.quarkus.vertx.runtime.VertxProducer_ProducerMethod_mutiny_d5befbd244a8a884fd08fff108d174c7e738c2d3_Bean.get(VertxProducer_ProducerMethod_mutiny_d5befbd244a8a884fd08fff108d174c7e738c2d3_Bean.zig:236)
        at io.quarkus.vault.runtime.client.VertxVaultClient_Bean.create(VertxVaultClient_Bean.zig:290)
        ... 65 more
```

And as we have seen in https://github.com/quarkusio/quarkus/pull/13537#issuecomment-755189920 and https://github.com/quarkusio/quarkus/pull/14142, synthetic beans cannot be accessed during the bootstrap configuration step.

vsevel commented:
@geoand you are fast ;)
is it doomed to fail then? or is there a trick we can use to make it work (I don't know, working with vertx at a lower level?)

geoand commented:
> @geoand you are fast ;)
> is it doomed to fail then? or is there a trick we can use to make it work (I don't know, working with vertx at a lower level?)

I don't know if it's doomed to fail - I'm just offering some caution :).

@cescoffier would know more

cescoffier commented:
Just for me to understand better.
Is the Vault extension use to source properties used in Quarkus config?

geoand commented:
@cescoffier exactly.

It is run in runtime-init, but before **any** of the other Quarkus recorders and returns a set of MP `ConfigSource` objects that are then used when creating the final MP Config object.
The reason it is run in runtime-init, is because Vault itself needs to be configured with some MP config properties (for example the URL of the vault server URL)

cescoffier commented:
Thanks,
So I would recommend creating a separate Vert.x instance (`Vertx.vertx()`), retrieving the configuration and closing the instance once done.

vsevel commented:
thanks @cescoffier 
I will get started with that approach.
stating the obvious: we will not be able to reuse the work done in other extensions such as `quarkus-vertx`, and we will not be able to leverage an existing `vertx` instance. as such it is falling short to this https://github.com/quarkusio/quarkus/pull/2897#issuecomment-503900843
however, this will allow to drop the `okhttp` client, which has been problematic in some cases, such as https://github.com/quarkusio/quarkus/issues/12680. plus `vertx` seems to be more _natural_ into `quarkus` than `okhttp`, making it a better long term solution.
so, if everybody is still fine, I will push forward this issue, and will try to reuse your code referenced in this https://github.com/quarkusio/quarkus/pull/8885#issuecomment-621013761
cc @geoand @sberyozkin

geoand commented:
Makes sense to me. 

As long as it's clear that in this case, Vert.x is an implementation detail and it is not expected that the various configuration settings will work on this very short lived Vert.x instance

cescoffier commented:
Yes, here, you just want: `Vertx.vertx()` - no configuration, just a new instance for the WebClient. Just be sure you close everything afterward.

sberyozkin commented:
Hi @vsevel this is great stuff; FYI `quarkus-oidc-client`/`OidcClientRecorder` uses the Vertx instance passed to it to init the client, but I guess the situation is different with Vault.
One thing I wanted to propose when I saw your PR the first time - IMHO it can be a good opportunity to update the user facing user interfaces to have the methods returning `Uni` - the current interface methods can all have the default implementations done in terms of these new methods awaiting indefinitely (Clement does not like this approach :-) but these are just existing blocking methods :-) )

That would allow the Vault extension be part of the endpoint specific Uni flows...if it sounds of interest than please consider doing it - may be in a follow up PR
thanks

vsevel commented:
hello all
so I have done what we discussed, except recoding the client to use the json mapping capabilities of vertx at this point.
I feared this would make too many changes now, and felt that by achieving already the replacement of `okhttp` by `vertx`, the main goal of this PR had been met. and additional work could be done in follow up PRs.

I have made a few minor adjustments, squashed and rebased. build is in progress.
so if everybody is fine, I would like to stop here for this PR and get your review.

> this is great stuff

thanks @sberyozkin :)

> IMHO it can be a good opportunity to update the user facing Vault interfaces (VaultKVService etc) to have the methods returning Uni

it would break the compatibility, but sure why not. I am not sure how much this is really needed by users.
@geoand and @cescoffier you think also this is the right thing to do?

geoand commented:
Do we really need to break compatibility? 

Can't we provide a Reactive alternative in addition to what we have?

gsmet commented:
If it wasn't reactive and using `CompletionStage` already, that would have my preference too.

But for sure, it can be another PR.

cescoffier commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/VertxVaultClient.java:
```diff
@@ -377,143 +386,147 @@ public void deletePolicy(String token, String name) {
+    private <T> T exec(HttpRequest<Buffer> request, Object body, Class<T> resultClass, int expectedCode) {
+        try {
+            HttpResponse<Buffer> response = body == null
+                    ? request.sendAndAwait()
```
We may want to be a bit more careful here. 
This may wait.... forever.

I would do:
```
request.send()
  .await().atMost(Duration.ofSecond(a_sensible_and_configurable_amount);
```

cescoffier replied:
Note that await is forbidden when you are on the event loop / I/O thread

vsevel replied:
I thought we would be protected with the idle timeout:
```
        WebClientOptions options = new WebClientOptions()
                .setConnectTimeout((int) vaultBootstrapConfig.connectTimeout.getSeconds())
                .setIdleTimeout((int) vaultBootstrapConfig.readTimeout.getSeconds());
```

is this equivalent to a read timeout? I looked at the definition in the doc, and it seems different. what is the proper way to do a request timeout? should that be on the request (using `HttpClientRequest.setTimeout`?), or something like this:

```
    private <T> T exec(HttpRequest<Buffer> request, Object body, Class<T> resultClass, int expectedCode) {
        try {
            Uni<HttpResponse<Buffer>> uni = body == null 
                    ? request.send() 
                    : request.sendBuffer(Buffer.buffer(requestBody(body)));
            HttpResponse<Buffer> response = uni.await().atMost(vaultConfigHolder.getVaultBootstrapConfig().readTimeout);
```

vsevel replied:
done. I reused the read timeout, which is part of the config.

vsevel replied:
or should that be:
```
            HttpRequest<Buffer> httpRequest = request.timeout(getRequestTimeout().toMillis());
            Uni<HttpResponse<Buffer>> uni = body == null 
                    ? httpRequest.send()
                    : httpRequest.sendBuffer(Buffer.buffer(requestBody(body)));
            HttpResponse<Buffer> response = uni.await().indefinitely();
```
how can we timeout at the socket level?

cescoffier commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/VertxVaultClient.java:
```diff
@@ -377,143 +386,147 @@ public void deletePolicy(String token, String name) {
-            throw new VaultException(e);
-        }
+    private int exec(HttpRequest<Buffer> request) {
+        return request.sendAndAwait().statusCode();
```
Same issue here, I would add a timeout, just in case...

vsevel replied:
done. I reused the read timeout, which is part of the config.

cescoffier commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/VertxVaultClient.java:
```diff
@@ -377,143 +386,147 @@ public void deletePolicy(String token, String name) {
         String body = null;
         try {
-            body = response.body().string();
+            body = response.body().toString();
```
Is encoding / charset something to handle here?

vsevel replied:
to be honest I am not sure. I assumed `UTF-8`
I could not find some info in the vault doc.
if I curl vault, I get something such as:
```
$ curl -v http://localhost:8200/v1/sys/health
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8200 (#0)
> GET /v1/sys/health HTTP/1.1
> Host: localhost:8200
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: no-store
< Content-Type: application/json
< Date: Fri, 22 Jan 2021 11:53:00 GMT
< Content-Length: 294
< 
{"initialized":true,"sealed":false,"standby":false,"performance_standby":false,"replication_performance_mode":"disabled","replication_dr_mode":"disabled","server_time_utc":1611316380,"version":"1.5.4","cluster_name":"vault-cluster-6dcd7526","cluster_id":"2e1e111c-7ac6-2499-358a-cd7402808803"}
* Connection #0 to host localhost left intact
```

so no encoding. should we assume `UTF-8` then?

in the [vault tests](https://github.com/quarkusio/quarkus/blob/master/test-framework/vault/src/main/java/io/quarkus/vault/test/VaultTestExtension.java#L69), I had added a `euro` to try to detect encoding issues (the secret is written using the cli, and read from the api). If we had an issue with encoding, I think this should have failed.

do you think we can do better?

cescoffier replied:
That's good enough for now. If we find out that encoding needs to be handled or configured, we can do that later.

cescoffier commented on extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/VertxVaultClient.java:
```diff
@@ -74,40 +74,52 @@
 
 @Singleton
-public class OkHttpVaultClient implements VaultClient {
+public class VertxVaultClient implements VaultClient {
```
(In another PR)

We may be able to provide a reactive version (ReactiveVaultClient) which would produce `Uni`.
With this one, the imperative (current) version would only delegate and append `await().atMost(...)`.

cescoffier commented:
Looks pretty good. I've made 2 comments.

One is for this PR. `await()` may wait forever. Adding a reasonable timeout would be safer. 

The second one is about having a reactive VaultClient. It can be done in a second PR, and it's pretty straightforward. The current client cannot be used from the I/O thread (because of the await()).

sberyozkin commented:
Hi guys, I was not proposing to rename, just something like
```
a();
Uni<a>();
```
only...

sberyozkin commented:
@vsevel by the way, please check https://github.com/quarkusio/quarkus/issues/14480#issuecomment-765369300, you may need to do the same if you haven't already done it - some extra coding is needed to correctly set the client options

vsevel commented:
> @vsevel by the way, please check [#14480 (comment)](https://github.com/quarkusio/quarkus/issues/14480#issuecomment-765369300), you may need to do the same if you haven't already done it - some extra coding is needed to correctly set the client options

I am using the `getAbs`:
```
    private HttpRequest<Buffer> builder(String path) {
        return webClient.getAbs(getUrl(path).toString());
    }
```

I have tests for ssl. so I think we are fine.

sberyozkin commented:
@vsevel sure, that method is named `setHttpsClientOptions` but it is mainly does a proper host and port settings; but if all is good for your PR then it is fine

vsevel commented:
@sberyozkin to be more precise I followed this [paragraph](https://vertx.io/docs/vertx-web-client/java/#_using_https), which happens to set the host, port and seems to turn on ssl, and I do not have to parse the url as in this [PR](https://github.com/quarkusio/quarkus/pull/14521/files).
the only drawback is that I have to build the full url for all requests, as opposed to pushing this info in the options.
do you think this is ok?

vsevel commented:
> Hi guys, I was not proposing to rename, just something like
> 
> ```
> a();
> Uni<a>();
> ```
> 
> only...

ah. I thought you wanted to replace the original signature. understood.

gsmet commented on extensions/vault/runtime/pom.xml:
```diff
@@ -19,8 +19,8 @@
-            <groupId>com.squareup.okhttp3</groupId>
-            <artifactId>okhttp</artifactId>
+            <groupId>io.smallrye.reactive</groupId>
+            <artifactId>smallrye-mutiny-vertx-web-client</artifactId>
```
Just to be sure, this one is working in native without any needs for an extension?

cescoffier replied:
Yes, works flawlessly in native.

gsmet commented:
OK, merged, then. Thanks everyone!

vsevel commented:
thanks @cescoffier for your review, and @sberyozkin for bootstrapping this effort a few months ago, given how easy it was to migrate to the vertx web client.
I am happy that we have been able to resolve several weaknesses in this extension's implementation over the last few months.
